### PR TITLE
Use ToLowerInvariant Or ToUpperInvariant Analyzer

### PR DESCRIPTION
**Pull request by TKharaishvili**
Fixes https://github.com/dotnet/runtime/issues/43976

The first commit only includes the analyzer. Once the analyzer code gets approved, I'll also add the fixer.

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: `IInvocationOperation.TargetMethod` is guaranteed to be non-null.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: How does this rule play with https://github.com/dotnet/roslyn-analyzers/issues/416?

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: This rule only gets activated for `ToUpper` and `ToLower` calls that have no arguments.

The other one(the one you've linked to) gets activated for `ToLower` calls with one argument and `ToLowerInvariant` calls.

The only place where they cross paths is `ToLowerInvariant`. The former recommends using `ToLowerInvariant` instead of `ToLower`.  The latter recommends against using `ToLowerInvariant` but for different reasons.

Basically here's the flow:
`str.ToLower()` => this analyzer => `str.ToLowerInvariant()` => the other analyzer.

Is this a problem?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (TKharaishvili) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: @mavasani how do I deal with rule id? I don't know how you guys number these things.
The diagnostics are complaining about it and it's causing the builds to fail.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: > Basically here's the flow:
`str.ToLower()` => this analyzer => `str.ToLowerInvariant()` => the other analyzer.
Is this a problem?

Yes, we avoid conflict recommendations from analyzers within the repo, i.e. an analyzer should not recommend use `y` instead of `x`, when we know that another analyzer will recommend use `z` instead of `y`, especially if both these rules are turned on by default.

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: Ok, so how should we deal with this particular issue? 
That other analyzer doesn't have a fixer so I'm not even sure what it's meant to do.
Should we drop this issue altogether?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: Really? Cause I've seen it being checked in several places. Does that mean those checks should be removed too?

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: Yes, likely that guarantee was not present in the IOperation API prototype prior to the eventual shipping API. Roslyn test suite runs the [TestOperationVisitor](http://source.roslyn.io/#Roslyn.Test.Utilities/Compilation/TestOperationVisitor.cs,19) on every Roslyn compiler unit test, which has the [following](http://source.roslyn.io/#Roslyn.Test.Utilities/Compilation/TestOperationVisitor.cs,458) assertion: `Assert.NotNull(operation.TargetMethod);`

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: Understood, thanks for the detailed info. So how should we go about removing those redundant checks? I could offer PRs as I find them but I'm not sure if that's an efficient way of handling it. Any thoughts?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: We need to decide if this gets a new rule ID or can be combined with any existing rule ID. If it is a new rule ID, then you identify the category of the rule and pick the next available rule ID from the rule range by category: https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/DiagnosticCategoryAndIdRanges.txt

As a followup when the PR gets merged, you should also file a doc bug to add documentation for it. For example, for globalization rule, you will file a content bug [here](https://docs.microsoft.com/en-us/visualstudio/code-quality/globalization-warnings?view=vs-2017) and also a follow-up bug in the repo to update the analyzer hyperlink to the newly created doc page.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: Tagging @sharwell @jinujoseph - we should discuss about this rule at the design meeting. Sam/Jinu, how do we want to track this given the issue/PR is in roslyn-analyzers repo.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs]: You can submit a single PR to delete all null checks. If you are fully inspired to cleanup our entire IOperation usage null checks, you can also remove null checks for other APIs that are asserted to be non-null in the TestOperationVisitor. I'd let you make a call on the scope of your PR.

**Comment by Evangelink** [comment]
@TKharaishvili Do you plan to update/finish this PR?

**Comment by TKharaishvili** [comment]
@Evangelink I think there were some open questions that needed to be discussed. If you look at the 2nd of my 3 subdiscussions with mavasani above, in the last comment he tagged some people mentioning it.

As far as I know those questions were never addressed.

At this point I don't really remember much about this issue or even the PR itself as it all happened over a year ago.

However I'm open to getting myself back up to speed on all of it given those open questions get addressed.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:43]: @TKharaishvili If you want to do so, please do it as a separate PR from this one.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:12]: Ping @sharwell @jinujoseph, could you please give us some feedback on this so that the PR can move forward?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:15]: @mavasani Shall it be a new rule ID?

**Comment by carlossanlop** [comment]
The proposed analyzer has not yet been approved. I'll move that issue to the runtime repo so it goes through API review. Once approved, we can continue with this PR review.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (carlossanlop) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:12]: @sharwell @jinujoseph we need your feedback here.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (carlossanlop) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:15]: @mavasani I don't think we need to get this analyzer merged with an existing rule. Let's use a brand new rule ID.

@TKharaishvili you can look at other merged PRs for guidance. For example, [here is a PR I recently merged](https://github.com/dotnet/roslyn-analyzers/pull/3641/files). Search for any instances of `CA2016` so you know which files you need to modify and you can replicate the same here.

**Review by sharwell — CHANGES_REQUESTED** [review]
I have two problems with the current design:

1. As an analyzer in the globalization category, rather than focus on usage of ToLowerInvariant and ToUpperInvariant, we should focus on explicitly indicating the desired culture for the conversion.
2. _Two_ code fixes should be offered. The first (primary) code fix should retain the call to `ToLower` but add an argument specifying current culture (semantic preservation). A second code fix can be offered to change the call to `ToLowerInvariant`.

If we keep the current approach of preferring to push users to change semantics by adopting invariant culture, the analyzer should probably be moved to the Performance category instead of the Globalization category, at which point the value seems to be diminished significantly.

**Comment by TKharaishvili** [comment]
@carlossanlop @tarekgh I managed to update the PR and address some of the original change requests. Can you gentlemen review it?
It only contains an Analyzer for now I'll start working on a Fixer soon.
Also the builds are failing with an error message that is a little hard to understand. I'd appreciate some help with that as well.

File: src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md
Comment (tarekgh) [inline src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:7]: Looks the link https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1311 is invalid link and maybe this is the cause of the CI failures you are seeing.

Comment (tarekgh) [reply src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:7]: @carlossanlop @buyaa-n could you please help with this error and have a look at the changes? I am seeing the changes are good to me.

Comment (TKharaishvili) [reply src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:7]: Thanks for replying so quickly. The link is invalid, that's true. But that makes sense for now doesn't it? I mean the analyzer itself doesn't exist yet so neither does the documentation...

Comment (mavasani) [reply src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:7]: @TKharaishvili is correct. The link doesn't need to be valid (in fact will be invalid when the PR is adding a new analyzer which will be documented in future). This is not the reason for the CI failure.

**Comment by TKharaishvili** [comment]
@carlossanlop @tarekgh gentlemen, can I request another round of code review please? Both the analyzer and fixer are implemented and ready to be looked at.
I see that the CI build is failing, I looked at the logs but I wasn't able to figure out what the cause might be.

**Review by tarekgh — APPROVED** [review]
LGTM, thanks @TKharaishvili for getting this done.

**Comment by tarekgh** [comment]
@carlossanlop @buyaa-n the changes look good to me. Could you have a look and merge it if you have permission to merge?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs:20]: We should use marker in the source for diagnostic locations instead of disabling this warning, will an added example in below test

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs:51]:
```suggestion
        string a = ""test"";
        [|a.ToLower()|];
        [|a?.ToLower()|];
    }
}");
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariantTests.cs:106]: The could be tests showing it would not flag other scenarios like `ToUpper()` used with a `Culture` option

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:28]: Not sure why it should be true
```suggestion
            isPortedFxCopRule: false,
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariant.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseToLowerInvariantOrToUpperInvariant.cs:24]: We need explicit message, having a bit more info than the title

**Comment by buyaa-n** [comment]
> I see that the CI build is failing, I looked at the logs but I wasn't able to figure out what the cause might be.

it seems you need to run `build -pack` to generate missing autogenerated files, please merge conflict and run the command

> can I request another round of code review please? Both the analyzer and fixer are implemented and ready to be looked at.

Very sorry for delayed review, it seems @sharwell's the requested change is not addressed, specifically the part: `Two code fixes should be offered. The first (primary) code fix should retain the call to ToLower but add an argument specifying current culture (semantic preservation). A second code fix can be offered to change the call to ToLowerInvariant.` Could you work on that? I also left few more comments

**Comment by buyaa-n** [comment]
> A question: as I understand @sharwell's request is not as much a code review as it is a request to change the actual task as it is described in the issue here - https://github.com/dotnet/runtime/issues/43976
> It asks to change quite a lot, it suggests to make the current fixer a secondary fix and introduce another fixer that would be primary. Since it is quite a big change I want to double-check, are we sure we want to implement this?

`string.ToUpper()` and `string.ToLower()` would use current culture by default, when fixing this with `string.ToUpperInvariant()`, `string.ToLowerInvariant()` users might be not aware of that the culture is changed to Invariant culture. So i support @sharwell's comment that we should not `push users to change semantics by adopting invariant culture`. I think this was also discussed in the API review some level (not the two fixer part though). I think suggesting 2 fixer would be perfect for this scenario, what you think about it @tarekgh?

> Also would you be able to point me to a similar analyzer that offers two code fixes at the same time so that I'm able to see how naming and all of that stuff should work in that case?

I believe 2 fixers are supported, but never used one myself, and don't know which analyzer does that, tagging @sharwell @mavasani  for help with that 

> The rest of the requested changes look much easier to make. I'll proceed with everything once I have the final answer to this.

Sure, thanks!

**Comment by tarekgh** [comment]
I agree with @sharwell comment about ``` Two code fixes should be offered. The first (primary) code fix should retain the call to ToLower but add an argument specifying current culture (semantic preservation). A second code fix can be offered to change the call to ToLowerInvariant.```.  Analyzers should offer these 2 options to the users.

**Comment by TKharaishvili** [comment]
@tarekgh @buyaa-n I addressed the latest change requests. Had to rename Analyzer, Fixer and Test classes since the old name didn't work for new requirements.

I hope no more major changes will be necessary.

Eager to hear your reply.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpper.cs
Comment (tarekgh) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpper.cs:56]: You can just use `==` operator instead of string.Equal here.

**Review by tarekgh — APPROVED** [review]
Thanks @TKharaishvili for getting this ready. LGTM. I think @buyaa-n is going to take a look too.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1862]: I believe our intended fix with this analysis is to use an invariant version, so we need to include that in the message somehow, something like: `Use invariant version or specify a culture to avoid implicit dependence on current culture`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1859]: I would like to mix it with old description 
```suggestion
    <value>Using an invariant version yields consistent results regardless of the culture of an application. Specify culture to help avoid accidental implicit dependence on current culture.</value>
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1856]:
```suggestion
    <value>Use invariant version or specify culture explicitly</value>
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs:51]: Is this index express the order of the fixer options? Is it possible to show the `UseInvariantVersion` option before `SpecifyCurrentCulture`?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.Fixer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.Fixer.cs:48]: NIT: preferably use is clause whenever possible
```suggestion
                if (node.Parent is MemberAccessExpressionSyntax memberAccess)
                {
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.Fixer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.Fixer.cs:55]: Same here

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.cs
Comment (buyaa-n) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpSpecifyCultureForToLowerAndToUpper.cs:27]: NIT:
```suggestion
            if (invocation.Expression is MemberAccessExpressionSyntax memberAccess)
            {
                return memberAccess.Name.GetLocation();
            }
            else if (invocation.Expression is MemberBindingExpressionSyntax memberBinding)
            {
                return memberBinding.Name.GetLocation();
            }
```

**Review by buyaa-n — APPROVED** [review]
Thanks @TKharaishvili I think the messaging needs some update as now it looks like an analyzer for suggesting to Specify `CurrentCulture`, and left a few optional NIT, other than that it looks good to me

**Comment by buyaa-n** [comment]
Tagging @sharwell for review as his requested change is addressed

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs:51]: oh I forgot the `SpecifyCurrentCulture` is requested to be first one, never mind

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx
Comment (TKharaishvili) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1856]: @buyaa-n To me `Use invariant version or specify current culture` implies that there are only those two options when it's actually possible to specify any culture. Would `Specify a culture or use an invariant version` be more suitable?

Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1856]: > Would `Specify a culture or use an invariant version` be more suitable?

Yes that sounds better, thanks

**Comment by TKharaishvili** [comment]
@tarekgh @buyaa-n  all the requested changes are made.

**Comment by buyaa-n** [comment]
Thank you @TKharaishvili, looks great to me, normally the reviewer who requested a change should approve or remove the CR before merge, tough I see @sharwell is out. As the requested changes addressed as he wanted hope he would not mind if I dismiss his change request to unblock the PR. We he have any other comment we can address that separately

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1868]: Looks like a typo? @buyaa-n

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpper.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpper.cs:49]: `OperationKind.ObjectCreation` is missing.

File: src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicSpecifyCultureForToLowerAndToUpper.vb
Comment (Youssef1313) [inline src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicSpecifyCultureForToLowerAndToUpper.vb:17]: This can fail I think. VB doesn't require parentheses for parameterless method calls

```vb
Public Class C
    Public Sub M(x As String)
        Dim y = x.ToLower ' This is a member access expression, not an invocation.
    End Sub
End Class
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyCultureForToLowerAndToUpperTests.Fixer.cs:14]: Tests should be consolidated into one file.

**Pull request merged**
